### PR TITLE
Improve RBS service deletion

### PR DESCRIPTION
**Pull request by cezarygerard**
In L4 RBS load balancer controller

1) remove service annotations, for  correct XLB->ILB transition when customer changes service annotations

2) try to delete the instance group to cleanup the instance group, this will
  - succeed when the last Service/LB using the instance group is deleted
  - fail if other L4/L7 load balancer is using the instance group
otherwise the IG can lea

### CODE REVIEW

**Comment by bowei** [comment]
Can we add more details to the commit message regarding what this fixes

File: pkg/l4lb/l4netlbcontroller.go [hunk 1]
```diff
@@ -415,12 +415,31 @@ func (lc *L4NetLBController) garbageCollectRBSNetLB(key string, svc *v1.Service)
 		result.Error = fmt.Errorf("Failed to reset L4 External LoadBalancer status, err: %w", err)
 		return result
 	}
+
+	// Also remove any LB annotations from the service metadata
```
Comment (bowei) [inline pkg/l4lb/l4netlbcontroller.go:419]: // Remove LB annotations from the Service when processing the finalizer.

File: pkg/l4lb/l4netlbcontroller.go [hunk 2]
```diff
@@ -415,12 +415,31 @@ func (lc *L4NetLBController) garbageCollectRBSNetLB(key string, svc *v1.Service)
...
+	if err := updateAnnotations(lc.ctx, svc, nil); err != nil {
+		lc.ctx.Recorder(svc.Namespace).Eventf(svc, v1.EventTypeWarning, "DeleteLoadBalancer",
+			"Error resetting resource annotations for load balancer: %v", err)
+		result.Error = fmt.Errorf("failed to reset resource annotations, err: %w", err)
```
Comment (bowei) [inline pkg/l4lb/l4netlbcontroller.go:423]: If we bail out here, the deletion will be retried?

File: pkg/l4lb/l4netlbcontroller.go [hunk 3]
```diff
@@ -415,12 +415,31 @@ func (lc *L4NetLBController) garbageCollectRBSNetLB(key string, svc *v1.Service)
 		result.Error = fmt.Errorf("Failed to reset L4 External LoadBalancer status, err: %w", err)
 		return result
 	}
+
+	// Also remove any LB annotations from the service metadata
+	if err := updateAnnotations(lc.ctx, svc, nil); err != nil {
+		lc.ctx.Recorder(svc.Namespace).Eventf(svc, v1.EventTypeWarning, "DeleteLoadBalancer",
+			"Error resetting resource annotations for load balancer: %v", err)
```
Comment (bowei) [inline pkg/l4lb/l4netlbcontroller.go:422]: Error removing resource annotations: %v

File: pkg/l4lb/l4netlbcontroller.go (hunk 2 above)
Comment (cezarygerard) [reply pkg/l4lb/l4netlbcontroller.go:423]: yes, the deletion will be retried.


this line will make it retry deletion if annotations are still there:
https://github.com/kubernetes/ingress-gce/blob/master/pkg/l4lb/l4netlbcontroller.go#L240

File: pkg/l4lb/l4netlbcontroller.go (hunk 3 above)
Comment (cezarygerard) [reply pkg/l4lb/l4netlbcontroller.go]: done

File: pkg/l4lb/l4netlbcontroller.go [hunk 4]
```diff
@@ -403,24 +403,45 @@ func (lc *L4NetLBController) garbageCollectRBSNetLB(key string, svc *v1.Service)
...
 	if err := common.EnsureDeleteServiceFinalizer(svc, common.NetLBFinalizerV2, lc.ctx.KubeClient); err != nil {
 		lc.ctx.Recorder(svc.Namespace).Eventf(svc, v1.EventTypeWarning, "DeleteLoadBalancerFailed",
 			"Error removing finalizer from L4 External LoadBalancer, err: %v", err)
 		result.Error = fmt.Errorf("Failed to remove L4 External LoadBalancer finalizer, err: %w", err)
 		return result
 	}
+
+	// Try to delete instance group, instancePool.DeleteInstanceGroup ignores errors if resource is in use or not found.
+	// TODO(cezarygerard) replace with multi-IG management
+	if err := lc.instancePool.DeleteInstanceGroup(lc.namer.InstanceGroup()); err != nil {
```
Comment (bowei) [inline pkg/l4lb/l4netlbcontroller.go:438]: can you add more information in the commit message why we need to add this extra deletion?

Under what conditions would this make a difference?

File: pkg/l4lb/l4netlbcontroller_test.go [hunk 5]
```diff
@@ -61,6 +62,11 @@ const (
 	userAddrName         = "UserStaticAddress"
 )
 
+var (
+	serviceAnnotationKeys = []string{annotations.FirewallRuleKey, annotations.BackendServiceKey, annotations.HealthcheckKey,
```
Comment (bowei) [inline pkg/l4lb/l4netlbcontroller_test.go:66]: break lines for easier maintenance

```
serviceAnnotationKeys = []string{
  annotations.FirewallRuleKey,
 ...
}
```

File: pkg/l4lb/l4netlbcontroller.go (hunk 4 above)
Comment (cezarygerard) [reply pkg/l4lb/l4netlbcontroller.go:438]: done

File: pkg/l4lb/l4netlbcontroller_test.go (hunk 5 above)
Comment (cezarygerard) [reply pkg/l4lb/l4netlbcontroller_test.go]: done

**Comment by cezarygerard** [comment]
@bowei 
all resolved

**Review by freehan — APPROVED** [review]
/lgtm
/approve

**Pull request merged**